src/aui/auibook.cpp

```diff
@@ -3446,6 +3446,140 @@ bool wxAuiNotebook::InsertPage(size_t index, wxWindow *page,
+            mergeHorizontal = true;
+        else
+            mergeHorizontal = lo2.m_dir >= 3;
+
+        if (mergeHorizontal)
+        {
+            m_size.x += lo2.m_size.x;
+            if (lo2.m_size.y > m_size.y) m_size.y = lo2.m_size.y;
+        }
+        else
+        {
+            if (lo2.m_size.x > m_size.x) m_size.x = lo2.m_size.x;
+            m_size.y += lo2.m_size.y;
+        }
+    }
+
+    wxSize m_size;
+    const wxAuiPaneInfo &m_pInfo;
+    unsigned char m_dir;
+
+    /* As the caulculation is done from the inner to the outermost pane, the
+     * panes are sorted in the following order: layer, direction, row,
+     * position. */
+    bool operator<(const wxAuiLayoutObject& lo2) const {
+        int diff = m_pInfo.dock_layer - lo2.m_pInfo.dock_layer;
+        if (diff)
+            return diff < 0;
+        diff = m_dir - lo2.m_dir;
+        if (diff)
+            return diff < 0;
+        diff = m_pInfo.dock_row - lo2.m_pInfo.dock_row;
+        if (diff)
+            return diff < 0;
+        return m_pInfo.dock_pos < lo2.m_pInfo.dock_pos;
+    }
+};
+
+wxSize wxAuiNotebook::DoGetBestSize() const
+{
+    /* The best size of the wxAuiNotebook is a combination of all panes inside
+     * the object. To be able to efficiently  calculate the dimensions (i.e.
+     * without iterating over the panes multiple times) the panes need to be
+     * processed in a specific order. Therefore we need to collect them in the
+     * following variable which is sorted later on. */
+    wxVector<wxAuiLayoutObject> layouts;
+    const wxAuiPaneInfoArray& all_panes =
+        const_cast<wxAuiManager&>(m_mgr).GetAllPanes();
+    const size_t pane_count = all_panes.GetCount();
+    const int tabHeight = GetTabCtrlHeight();
+    for (size_t n = 0; n < pane_count; ++n)
+    {
+        const wxAuiPaneInfo &pInfo = all_panes[n];
+        if (pInfo.name == wxT("dummy") || pInfo.IsFloating())
+            continue;
+
+        const wxTabFrame* tabframe = (wxTabFrame*) all_panes.Item(n).window;
+        const wxAuiNotebookPageArray &pages = tabframe->m_tabs->GetPages();
+
+        wxSize bestPageSize;
+        for (size_t pIdx = 0; pIdx < pages.GetCount(); pIdx++)
+            bestPageSize.IncTo(pages[pIdx].window->GetBestSize());
+
+        bestPageSize.y += tabHeight;
+        // Store the current pane with its largest window dimensions
+        layouts.push_back(wxAuiLayoutObject(bestPageSize, pInfo));
+    }
+    wxVectorSort(layouts);
+
+    /* The sizes of the panes are merged here. As the center pane is always at
+     * position 0 all sizes are merged there. As panes can be stacked using the
+     * dock_pos property, different positions are merged at the first (i.e.
+     * dock_pos = 0) element before being merged with the center pane. */
+    size_t pos = 0;
+    for (size_t n = 1; n < layouts.size(); n++)
+    {
+        if (layouts[n].m_pInfo.dock_layer != layouts[pos].m_pInfo.dock_layer)
```

Comment:
The bodies of all 3 `if`s are identical, so why can't we simplify this code to (in pseudo-code)
```
if ( SameUpToPos(layouts[n].m_pInfo, layouts[pos].m_pInfo) )
{
    layouts[pos].MergeLayout(layouts[n]);
}
else
{
    layouts[0].MergeLayout(layouts[pos]);
    pos = n;
}
```

Am I missing something here?

Reply:
No, your right. I simply programmed that from top to bottom and didn't recognize the equal cases.